File: python/src/contriwork_market_data/adapters/polygon_io.py
"""Polygon.io adapter — stocks_us + forex.

Endpoints used:
- ``/v2/last/trade/{ticker}`` — latest trade as spot.
- ``/v2/aggs/ticker/{ticker}/range/{multiplier}/{timespan}/{from}/{to}`` —
  historical aggregate bars.

Auth: ``apiKey`` query parameter. Free tier: 5 req/min, 2-year history.
"""

from __future__ import annotations

from collections.abc import AsyncIterator, Awaitable, Callable
from datetime import UTC, datetime
from decimal import Decimal
from typing import Any

import httpx

from .._adapter_helpers import resolve_api_key, streaming_not_supported
from .._http import build_async_client, request_json
from ..errors import (
    AdapterFeatureNotSupportedError,
    AdapterUnavailableError,
    InvalidIntervalError,
    SymbolNotFoundError,
)
from ..types import (
    Candle,
    Capability,
    Interval,
    OrderBook,
    SpotPrice,
    Ticker,
)

__all__ = ["PolygonIOAdapter"]


_DEFAULT_BASE = "https://api.polygon.io"
_RATE_LIMIT_PER_MINUTE = 5

# Polygon aggs use (multiplier, timespan). Map our Interval enum to (m, span).
_INTERVAL_MAP: dict[Interval, tuple[int, str]] = {
    Interval.M1: (1, "minute"),
    Interval.M5: (5, "minute"),
    Interval.M15: (15, "minute"),
    Interval.M30: (30, "minute"),
    Interval.H1: (1, "hour"),
    Interval.H4: (4, "hour"),
    Interval.D1: (1, "day"),
    Interval.W1: (1, "week"),
    Interval.MN1: (1, "month"),
}

_SUPPORTED_INTERVALS: tuple[Interval, ...] = tuple(_INTERVAL_MAP.keys())


class PolygonIOAdapter:
    adapter_id = "polygon-io"

    def __init__(
        self,
        *,
        api_key: str | None = None,
        api_key_provider: Callable[[], Awaitable[str | None]] | None = None,
        base_url: str | None = None,
        timeout_s: float = 15.0,
        http_proxy: str | None = None,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        self._api_key = api_key
        self._api_key_provider = api_key_provider
        self._base_url = (base_url or _DEFAULT_BASE).rstrip("/")
        self._timeout_s = timeout_s
        self._http_proxy = http_proxy
        self._client = http_client
        self._owns_client = http_client is None
        self.capability = Capability(
            supported_markets=("stocks_us", "forex"),
            supported_intervals=_SUPPORTED_INTERVALS,
            supported_quote_currencies=("USD",),
            supports_order_book=False,
            supports_native_streaming=False,
            rate_limit_per_minute=_RATE_LIMIT_PER_MINUTE,
            requires_auth=True,
        )

    async def aclose(self) -> None:
        if self._owns_client and self._client is not None:
            await self._client.aclose()
            self._client = None

    async def _ensure_client(self) -> httpx.AsyncClient:
        if self._client is None:
            self._client = build_async_client(
                timeout_s=self._timeout_s,
                http_proxy=self._http_proxy,
            )
        return self._client

    async def _api_key_value(self) -> str:
        key = await resolve_api_key(
            adapter_id=self.adapter_id,
            api_key=self._api_key,
            api_key_provider=self._api_key_provider,
            required=True,
        )
        assert key is not None
        return key

    async def get_spot(self, symbol: str, quote_currency: str) -> SpotPrice:
        client = await self._ensure_client()
        key = await self._api_key_value()
        try:
            payload = await request_json(
                client,
                "GET",
                f"{self._base_url}/v2/last/trade/{symbol}",
                adapter_id=self.adapter_id,
                params={"apiKey": key},
            )
        except AdapterUnavailableError as exc:
            if "HTTP 404" in str(exc):
                raise SymbolNotFoundError(
                    f"polygon-io does not know ticker {symbol!r}",
                    adapter_id=self.adapter_id,
                ) from exc
            raise
        if not isinstance(payload, dict) or "results" not in payload:
            raise AdapterUnavailableError(
                "polygon-io trade returned unexpected payload",
                adapter_id=self.adapter_id,
            )
        results = payload["results"]
        price = _opt_decimal(results.get("p"))
        if price is None:
            raise SymbolNotFoundError(
                f"polygon-io returned no last trade for {symbol!r}",
                adapter_id=self.adapter_id,
            )
        # Polygon timestamps are in nanoseconds.
        ts_ns = int(results.get("t", 0)) or 0
        return SpotPrice(
            symbol=symbol,
            last=price,
            quote_currency=quote_currency,
            timestamp=datetime.fromtimestamp(ts_ns / 1e9, tz=UTC),
            source_adapter=self.adapter_id,
            volume_24h=_opt_decimal(results.get("s")),
        )

    async def get_ohlcv(
        self,
        symbol: str,
        interval: Interval,
        since: datetime | None,
        limit: int,
    ) -> list[Candle]:
        if interval not in _SUPPORTED_INTERVALS:
            raise InvalidIntervalError(
                f"polygon-io does not support interval {interval.value}",
                adapter_id=self.adapter_id,
            )
        client = await self._ensure_client()
        key = await self._api_key_value()
        multiplier, span = _INTERVAL_MAP[interval]
        # Default lookback: 30 days when no since is supplied.
        end = datetime.now(UTC)
        start = since or end.replace(year=max(1970, end.year - 1))
        try:
            payload = await request_json(
                client,
                "GET",
                (
                    f"{self._base_url}/v2/aggs/ticker/{symbol}/range/"
                    f"{multiplier}/{span}/{start:%Y-%m-%d}/{end:%Y-%m-%d}"
                ),
                adapter_id=self.adapter_id,
                params={"apiKey": key, "limit": min(limit, 5000)},
            )
        except AdapterUnavailableError as exc:
            if "HTTP 404" in str(exc):
                raise SymbolNotFoundError(
                    f"polygon-io does not know ticker {symbol!r}",
                    adapter_id=self.adapter_id,
                ) from exc
            raise
        if not isinstance(payload, dict):
            raise AdapterUnavailableError(
                "polygon-io aggs returned unexpected payload",
                adapter_id=self.adapter_id,
            )
        results = payload.get("results") or []
        candles: list[Candle] = []
        for row in results[:limit]:
            ts_ms = int(row.get("t", 0))
            candles.append(
                Candle(
                    timestamp=datetime.fromtimestamp(ts_ms / 1000, tz=UTC),
                    open=Decimal(str(row["o"])),
                    high=Decimal(str(row["h"])),
                    low=Decimal(str(row["l"])),
                    close=Decimal(str(row["c"])),
                    volume=Decimal(str(row.get("v", 0))),
                    trade_count=row.get("n"),
                )
            )
        return candles

    async def get_order_book(self, symbol: str, depth: int) -> OrderBook:
        raise AdapterFeatureNotSupportedError(
            "polygon-io order book requires the L2 paid tier and is out of v0.1.0 scope",
            adapter_id=self.adapter_id,
        )

    def subscribe_ticker(self, symbol: str) -> AsyncIterator[Ticker]:
        return streaming_not_supported(self.adapter_id)


def _opt_decimal(value: Any) -> Decimal | None:
    if value is None or value == "":
        return None
    return Decimal(str(value))

File: typescript/src/adapters/polygon-io.ts
/**
 * Polygon.io adapter — US stocks + forex. `/v2/last/trade/{ticker}` +
 * aggregates. Free tier: 5 req/min.
 */
import { Decimal } from "decimal.js";
import type { ApiKeyProvider, MarketDataAdapter } from "../adapter.js";
import { resolveApiKey } from "../internal/credentials.js";
import {
  type FetchLike,
  defaultFetch,
  getJson,
  queryString,
} from "../internal/http.js";
import { streamingNotSupported } from "../internal/streaming.js";
import {
  AdapterFeatureNotSupportedError,
  AdapterUnavailableError,
  InvalidIntervalError,
  SymbolNotFoundError,
} from "../errors.js";
import {
  type Candle,
  type Capability,
  EMPTY_EXTRA,
  Interval,
  type OrderBook,
  type SpotPrice,
  type Ticker,
} from "../types.js";

const INTERVAL_MAP = new Map<Interval, { multiplier: number; span: string }>([
  [Interval.M1, { multiplier: 1, span: "minute" }],
  [Interval.M5, { multiplier: 5, span: "minute" }],
  [Interval.M15, { multiplier: 15, span: "minute" }],
  [Interval.M30, { multiplier: 30, span: "minute" }],
  [Interval.H1, { multiplier: 1, span: "hour" }],
  [Interval.H4, { multiplier: 4, span: "hour" }],
  [Interval.D1, { multiplier: 1, span: "day" }],
  [Interval.W1, { multiplier: 1, span: "week" }],
  [Interval.MN1, { multiplier: 1, span: "month" }],
]);

/** Options for {@link PolygonIOAdapter}. */
export interface PolygonIOOptions {
  readonly apiKey?: string;
  readonly apiKeyProvider?: ApiKeyProvider;
  readonly baseUrl?: string;
  readonly fetchFn?: FetchLike;
}

function ymd(date: Date): string {
  return date.toISOString().slice(0, 10);
}

/** Polygon.io stocks + forex adapter. */
export class PolygonIOAdapter implements MarketDataAdapter {
  public readonly adapterId = "polygon-io";
  public readonly capability: Capability;

  private readonly apiKey?: string;
  private readonly apiKeyProvider?: ApiKeyProvider;
  private readonly baseUrl: string;
  private readonly fetchFn: FetchLike;

  public constructor(options: PolygonIOOptions = {}) {
    if (options.apiKey !== undefined) {
      this.apiKey = options.apiKey;
    }
    if (options.apiKeyProvider !== undefined) {
      this.apiKeyProvider = options.apiKeyProvider;
    }
    this.baseUrl = (options.baseUrl ?? "https://api.polygon.io").replace(
      /\/+$/,
      "",
    );
    this.fetchFn = options.fetchFn ?? defaultFetch;
    this.capability = {
      supportedMarkets: ["stocks_us", "forex"],
      supportedIntervals: [...INTERVAL_MAP.keys()],
      supportedQuoteCurrencies: ["USD"],
      supportsOrderBook: false,
      supportsNativeStreaming: false,
      rateLimitPerMinute: 5,
      requiresAuth: true,
    };
  }

  private async key(signal?: AbortSignal): Promise<string> {
    return (await resolveApiKey(
      this.adapterId,
      this.apiKey,
      this.apiKeyProvider,
      true,
      signal,
    )) as string;
  }

  private async get(
    path: string,
    query: Record<string, string | number | undefined>,
    signal?: AbortSignal,
  ): Promise<unknown> {
    try {
      return await getJson(
        this.fetchFn,
        this.adapterId,
        this.baseUrl + path + queryString(query),
        undefined,
        signal,
      );
    } catch (err) {
      if (
        err instanceof AdapterUnavailableError &&
        err.message.includes("HTTP 404")
      ) {
        throw new SymbolNotFoundError(
          `polygon-io does not know ticker (404): ${path}`,
          this.adapterId,
        );
      }
      throw err;
    }
  }

  public async getSpot(
    symbol: string,
    quoteCurrency: string,
    signal?: AbortSignal,
  ): Promise<SpotPrice> {
    const key = await this.key(signal);
    const root = (await this.get(
      `/v2/last/trade/${symbol}`,
      { apiKey: key },
      signal,
    )) as Record<string, unknown>;
    const results = root["results"] as Record<string, unknown> | undefined;
    if (results === undefined) {
      throw new AdapterUnavailableError(
        "polygon-io trade returned unexpected payload",
        this.adapterId,
      );
    }
    if (results["p"] === undefined) {
      throw new SymbolNotFoundError(
        `polygon-io returned no last trade for '${symbol}'`,
        this.adapterId,
      );
    }
    const tsNs = Number(results["t"] ?? 0);
    const vol =
      results["s"] !== undefined
        ? new Decimal(results["s"] as Decimal.Value)
        : undefined;
    return {
      symbol,
      last: new Decimal(results["p"] as Decimal.Value),
      quoteCurrency,
      timestamp: new Date(tsNs / 1_000_000),
      sourceAdapter: this.adapterId,
      extra: EMPTY_EXTRA,
      ...(vol !== undefined && { volume24h: vol }),
    };
  }

  public async getOhlcv(
    symbol: string,
    interval: Interval,
    since: Date | undefined,
    limit: number,
    signal?: AbortSignal,
  ): Promise<readonly Candle[]> {
    const spec = INTERVAL_MAP.get(interval);
    if (spec === undefined) {
      throw new InvalidIntervalError(
        `polygon-io does not support interval ${interval}`,
        this.adapterId,
      );
    }
    const key = await this.key(signal);
    const end = new Date();
    const start = since ?? new Date(end.getTime() - 365 * 24 * 60 * 60 * 1000);
    const path = `/v2/aggs/ticker/${symbol}/range/${String(spec.multiplier)}/${spec.span}/${ymd(start)}/${ymd(end)}`;
    const root = (await this.get(
      path,
      { apiKey: key, limit: Math.min(limit, 5000) },
      signal,
    )) as Record<string, unknown>;
    const results =
      (root["results"] as Record<string, unknown>[] | undefined) ?? [];
    return results.slice(0, limit).map((row) => ({
      timestamp: new Date(Number(row["t"] ?? 0)),
      open: new Decimal(row["o"] as Decimal.Value),
      high: new Decimal(row["h"] as Decimal.Value),
      low: new Decimal(row["l"] as Decimal.Value),
      close: new Decimal(row["c"] as Decimal.Value),
      volume:
        row["v"] !== undefined
          ? new Decimal(row["v"] as Decimal.Value)
          : new Decimal(0),
      ...(row["n"] !== undefined ? { tradeCount: Number(row["n"]) } : {}),
      extra: EMPTY_EXTRA,
    }));
  }

  public getOrderBook(
    _symbol: string,
    _depth: number,
    _signal?: AbortSignal,
  ): Promise<OrderBook> {
    return Promise.reject(
      new AdapterFeatureNotSupportedError(
        "polygon-io order book requires the L2 paid tier and is out of v0.1.0 scope",
        this.adapterId,
      ),
    );
  }

  public subscribeTicker(
    _symbol: string,
    _signal?: AbortSignal,
  ): AsyncIterable<Ticker> {
    return streamingNotSupported(this.adapterId);
  }
}
